### PR TITLE
Checks information is being recovered from a json file

### PR DESCRIPTION
**Pull request by kevinTveizaga**
Rules for checks now are being activated from a json file

### CODE REVIEW

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/Remediation.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.fundacionjala.enforce.sonarqube.apex.rules;
+
+import org.sonar.api.server.debt.DebtRemediationFunction;
+import org.sonar.api.server.rule.RulesDefinition.DebtRemediationFunctions;
+
+public class Remediation {
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/Remediation.java:13]: missing doc

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/Remediation.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
+/*
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/Remediation.java:1]: invalid class header

File: sonar-apex-plugin/src/test/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinitionTest.java [hunk 3]
```diff
@@ -2,22 +2,35 @@
  * Copyright (c) Fundacion Jala. All rights reserved.
  * Licensed under the MIT license. See LICENSE file in the project root for full license information.
  */
-
 package org.fundacionjala.enforce.sonarqube.apex.rules;
 
+
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/test/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinitionTest.java]: delete the blank line

File: sonar-apex-plugin/src/test/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinitionTest.java [hunk 4]
```diff
@@ -2,22 +2,35 @@
...
 import static org.junit.Assert.assertThat;
 
 public class ApexRulesDefinitionTest {
 
+    private ApexRulesDefinition rulesDefinition;
+
+    @Before
+    public void setUp() {
+        rulesDefinition = new ApexRulesDefinition();
+
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/test/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinitionTest.java]: delete the blank line

File: apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1002.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    "func": "Constant\/Issue",
+    "constantCost": "1min"
+  },
+  "sqaleSubCharac": "READABILITY",
+  "tags": [
+    "convention"
+  ],
+  "defaultSeverity": "Minor"
+}
+
```
Comment (JimmyRojas) [inline apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1002.json]: delete the blank line

File: apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1003.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    "func": "Constant\/Issue",
+    "constantCost": "5min"
+  },
+  "sqaleSubCharac": "READABILITY",
+  "tags": [
+    "bug"
+  ],
+  "defaultSeverity": "Critical"
+}
+
```
Comment (JimmyRojas) [inline apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1003.json]: delete the blank line

File: apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1004.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    "func": "Constant\/Issue",
+    "constantCost": "5min"
+  },
+  "sqaleSubCharac": "READABILITY",
+  "tags": [
+    "bug"
+  ],
+  "defaultSeverity": "Critical"
+}
+
```
Comment (JimmyRojas) [inline apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1004.json]: delete the blank line

File: apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1005.json [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    "func": "Constant\/Issue",
+    "constantCost": "5min"
+  },
+  "sqaleSubCharac": "READABILITY",
+  "tags": [
+    "bug"
+  ],
+  "defaultSeverity": "Critical"
+}
+
```
Comment (JimmyRojas) [inline apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1005.json]: delete the blank line

File: apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1006.json [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    "func": "Constant\/Issue",
+    "constantCost": "1min"
+  },
+  "sqaleSubCharac": "READABILITY",
+  "tags": [
+    "obsolete"
+  ],
+  "defaultSeverity": "Info"
+}
+
```
Comment (JimmyRojas) [inline apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1006.json]: delete the blank line

File: apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1008.json [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+  "remediation": {
+    "func": "Constant\/Issue",
+    "constantCost": "6min"
+  },
+  "sqaleSubCharac": "READABILITY",
+  "tags": [
+    "convention"
+  ],
+  "defaultSeverity": "Minor"
+}
```
Comment (JimmyRojas) [inline apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1008.json]: add missing blank line

File: apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1009.json [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+    "func": "Constant\/Issue",
+    "constantCost": "5min"
+  },
+  "sqaleSubCharac": "SYNCHRONIZATION_RELIABILITY",
+  "tags": [
+    "obsolete"
+  ],
+  "defaultSeverity": "Info"
+}
+
```
Comment (JimmyRojas) [inline apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1009.json]: delete the blank line

File: apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1010.json [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+    "func": "Constant\/Issue",
+    "constantCost": "1min"
+  },
+  "sqaleSubCharac": "READABILITY",
+  "tags": [
+    "convention"
+  ],
+  "defaultSeverity": "Major"
+}
+
```
Comment (JimmyRojas) [inline apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1010.json]: delete the blank line

File: apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1021.json [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+    "func": "Constant\/Issue",
+    "constantCost": "5min"
+  },
+  "sqaleSubCharac": "MODULARITY",
+  "tags": [
+    "confusing"
+  ],
+  "defaultSeverity": "Major"
+}
+
```
Comment (JimmyRojas) [inline apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1021.json]: delete the blank line

File: apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1024.json [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+    "func": "Constant\/Issue",
+    "constantCost": "1min"
+  },
+  "sqaleSubCharac": "UNIT_TESTS",
+  "tags": [
+    "convention"
+  ],
+  "defaultSeverity": "Major"
+}
+
```
Comment (JimmyRojas) [inline apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1024.json]: delete the blank line

File: apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1025.json [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+    "func": "Constant\/Issue",
+    "constantCost": "1min"
+  },
+  "sqaleSubCharac": "LOGIC_RELIABILITY",
+  "tags": [
+    "convention"
+  ],
+  "defaultSeverity": "Major"
+}
+
```
Comment (JimmyRojas) [inline apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1025.json]: delete the blank line

File: apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1026.json [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+    "func": "Constant\/Issue",
+    "constantCost": "2min"
+  },
+  "sqaleSubCharac": "READABILITY",
+  "tags": [
+    "convention"
+  ],
+  "defaultSeverity": "Major"
+}
+
```
Comment (JimmyRojas) [inline apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1026.json]: delete the blank line

File: apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/LineLength.json [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+    "func": "Constant\/Issue",
+    "constantCost": "1min"
+  },
+  "sqaleSubCharac": "READABILITY",
+  "tags": [
+    "convention"
+  ],
+  "defaultSeverity": "Minor"
+}
+
```
Comment (JimmyRojas) [inline apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/LineLength.json]: delete the blank line

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java [hunk 18]
```diff
@@ -25,8 +43,73 @@ public void define(Context context) {
...
+            rule.setDebtSubCharacteristic(metadata.sqaleSubCharac);
+            if (metadata.remediation != null) {
+                rule.setDebtRemediationFunction(metadata.remediation.remediationFunction(rule.debtRemediationFunctions()));
+                rule.setEffortToFixDescription(metadata.remediation.linearDesc);
+            }
+        }
+    }
+
+    private static void addHtmlDescription(NewRule rule, String metadataKey) {
+        URL resource = ApexRulesDefinition.class.getResource(RESOURCE_BASE_PATH + "/" + metadataKey + ".html");
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java]: use a String template

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java [hunk 19]
```diff
@@ -25,8 +43,73 @@ public void define(Context context) {
...
+        if (rule == null) {
+            throw new IllegalStateException("No rule was created for " + ruleClass + " in " + repository.key());
+        }
+
+        addHtmlDescription(rule, rule.key());
+        addMetadata(rule, rule.key());
+    }
+
+    private void addMetadata(NewRule rule, String metadataKey) {
+        URL resource = ExternalDescriptionLoader.class.getResource(RESOURCE_BASE_PATH + "/" + metadataKey + ".json");
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java]: use a String template

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java [hunk 20]
```diff
@@ -25,8 +43,73 @@ public void define(Context context) {
...
+            newRule(ruleClass, repository);
+        });
         repository.done();
     }
+
+    @VisibleForTesting
+    protected void newRule(Class<?> ruleClass, NewRepository repository) {
+        org.sonar.check.Rule ruleAnnotation = AnnotationUtils.getAnnotation(ruleClass, org.sonar.check.Rule.class);
+        if (ruleAnnotation == null) {
+            throw new IllegalArgumentException("No Rule annotation was found on " + ruleClass);
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java]: use a String template

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java [hunk 21]
```diff
@@ -25,8 +43,73 @@ public void define(Context context) {
...
+        }
+        String ruleKey = ruleAnnotation.key();
+        if (StringUtils.isEmpty(ruleKey)) {
+            throw new IllegalArgumentException("No key is defined in Rule annotation of " + ruleClass);
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java]: use a String template

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java [hunk 22]
```diff
@@ -25,8 +43,73 @@ public void define(Context context) {
...
+        }
+        NewRule rule = repository.rule(ruleKey);
+        if (rule == null) {
+            throw new IllegalStateException("No rule was created for " + ruleClass + " in " + repository.key());
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java]: use a String template

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java [hunk 23]
```diff
@@ -25,8 +43,73 @@ public void define(Context context) {
...
+
+    private static String readResource(URL resource) {
+        try {
+            return Resources.toString(resource, Charsets.UTF_8);
+        } catch (IOException e) {
+            throw new IllegalStateException("Failed to read: " + resource, e);
+        }
+    }
+
+    private static class RuleMetatada {
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java:133]: missing doc

File: sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java [hunk 24]
```diff
@@ -25,8 +43,73 @@ public void define(Context context) {
...
+
+        String title;
+        String status;
+        @Nullable
+        Remediation remediation;
+
+        String sqaleSubCharac;
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/main/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinition.java]: can we be more descriptive regarding naming convention?

File: sonar-apex-plugin/src/test/java/org/fundacionjala/enforce/sonarqube/apex/rules/RemediationTest.java [hunk 25]
```diff
@@ -0,0 +1,85 @@
+/*
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/test/java/org/fundacionjala/enforce/sonarqube/apex/rules/RemediationTest.java:1]: invalid class header

File: sonar-apex-plugin/src/test/java/org/fundacionjala/enforce/sonarqube/apex/rules/RemediationTest.java [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.fundacionjala.enforce.sonarqube.apex.rules;
+
+
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/test/java/org/fundacionjala/enforce/sonarqube/apex/rules/RemediationTest.java]: delete the blank line

File: sonar-apex-plugin/pom.xml [hunk 27]
```diff
@@ -37,6 +37,26 @@
...
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
+            <artifactId>gson</artifactId>
+            <version>2.6.2</version>
+            <type>jar</type>
+        </dependency>
+        <dependency>
+            <groupId>org.easytesting</groupId>
```
Comment (JimmyRojas) [inline sonar-apex-plugin/pom.xml:47]: where are we using this library?

File: sonar-apex-plugin/pom.xml [hunk 28]
```diff
@@ -37,6 +37,26 @@
...
+            <artifactId>fest-assert</artifactId>
+            <version>1.4</version>
+            <scope>test</scope>
+            <type>jar</type>
+        </dependency>
+        <dependency>
+            <groupId>org.easytesting</groupId>
```
Comment (JimmyRojas) [inline sonar-apex-plugin/pom.xml:54]: where are we using this library?

**Comment by JimmyRojas** [comment]
please see a lot of comments above

File: sonar-apex-plugin/pom.xml (hunk 27 above)
Comment (kevinTveizaga) [inline sonar-apex-plugin/pom.xml:47]: ApexRulesDefinition line: 76  to "adapt" in this case to recover information from Json files

File: sonar-apex-plugin/pom.xml (hunk 28 above)
Comment (kevinTveizaga) [inline sonar-apex-plugin/pom.xml:54]: ApexrulesDefinitionTest for assertions of test_invalid_checks which btw is beeing renamed

File: sonar-apex-plugin/src/test/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinitionTest.java [hunk 29]
```diff
@@ -26,4 +36,50 @@ public void test() {
...
+                    .hasMessage(String.format("No key is defined in Rule annotation of class %s", EmptyRuleKey.class.getName()));
+        }
+
+        try {
+            rulesDefinition.newRule(UnregisteredRule.class, newRepository);
+        } catch (IllegalStateException illegalStateException) {
+            assertThat(illegalStateException)
+                    .hasMessage(String.format("No rule was created for class %s in test", UnregisteredRule.class.getName()));
+        }
+        // no metadata defined, does not fail on registration of rule
```
Comment (JimmyRojas) [inline sonar-apex-plugin/src/test/java/org/fundacionjala/enforce/sonarqube/apex/rules/ApexRulesDefinitionTest.java]: do not add inline comments

File: apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1026.json [hunk 30]
```diff
@@ -0,0 +1,13 @@
...
+    "remediation": {
+        "func": "Constant\/Issue",
+        "constantCost": "2min"
+    },
+    "sqaleSubCharacteristic": "READABILITY",
+    "tags": [
+        "convention"
+    ],
+    "defaultSeverity": "Major"
+}
```
Comment (JimmyRojas) [inline apex-checks/src/main/resources/org/sonar/l10n/apex/rules/apex/A1026.json]: a new line is missing

**Comment by JimmyRojas** [comment]
just 2 minor updates

**Pull request merged**